jobs/spFilteringJob.go

```diff
@@ -0,0 +1,135 @@
+package main
+
+import (
+	"context"
+	"log"
+	"os"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/data-preservation-programs/spade-tenant/config"
+	"github.com/data-preservation-programs/spade-tenant/db"
+	"github.com/joho/godotenv"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+type SP struct {
+	Version        int       `bson:"__v"`
+	Provider       string    `bson:"_id"`
+	AgentCity      string    `bson:"agentCity"`
+	AgentCountry   string    `bson:"agentCountry"`
+	AgentLatitude  float64   `bson:"agentLatitude"`
+	AgentLongitude float64   `bson:"agentLongitude"`
+	AgentRegion    string    `bson:"agentRegion"`
+	Date           time.Time `bson:"date"`
+	LatencyMs      float64   `bson:"latencyMs"`
+	Multiaddr      string    `bson:"multiaddr"`
+	TestId         string    `bson:"testId"`
+}
+
+func main() {
+	godotenv.Load()
+	serverAPI := options.ServerAPI(options.ServerAPIVersion1)
+	opts := options.Client().ApplyURI(os.Getenv("MONGODB_URI")).SetServerAPIOptions(serverAPI)
+
+	// Create a new client and connect to the server
+	client, err := mongo.Connect(context.TODO(), opts)
+	if err != nil {
+		panic(err)
+	}
+
+	defer func() {
+		if err = client.Disconnect(context.TODO()); err != nil {
+			panic(err)
+		}
+	}()
+
+	// Connect to collection with sp data
+	collection := client.Database(os.Getenv("MONGODB_DATABASE")).Collection(os.Getenv("MONGODB_SP_COLLECTION"))
+
+	cur, _ := collection.Find(context.TODO(), bson.D{{}})
+	var sps []SP
+
+	for cur.Next(context.TODO()) {
+		var sp SP
+		err := cur.Decode(&sp)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		sps = append(sps, sp)
+	}
+
+	var clauses []db.TenantSPEligibilityClauses
+	config := config.InitConfig()
+	database, err := db.OpenDatabase(config.DB_URL, config.DEBUG, config.DRY_RUN)
+
+	database.Model(&db.TenantSPEligibilityClauses{}).Find(&clauses)
+	tenantClauseMap := make(map[db.ID][]db.TenantSPEligibilityClauses)
+
+	for _, clause := range clauses {
+		if _, ok := tenantClauseMap[clause.TenantID]; !ok {
+			tenantClauseMap[clause.TenantID] = make([]db.TenantSPEligibilityClauses, 0)
+		}
+
+		tenantClauseMap[clause.TenantID] = append(tenantClauseMap[clause.TenantID], clause)
+	}
+
+	for tenantId, clauses := range tenantClauseMap {
+		for _, sp := range sps {
+			isValid := true
+
+			for _, clause := range clauses {
+				if !compare(clause.ClauseValue, getField(&sp, clause.ClauseAttribute), clause.ClauseOperator) {
```

Comment:
do you think it would be possible to write some unit tests for the business logic of this filtering job? 

The best would be if we can extract the logic (iteration/comparison) apart from the DB stuff into a separate function, and then run some unit tests against it for a variety of clauses/SPs. At minimum some unit tests of the `compare/contains/getField` would be good.

Reply:
Absolutely. This was more of a wip pr to flush out any issues with the model and business logic.